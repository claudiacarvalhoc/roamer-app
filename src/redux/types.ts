export const PROJECT_LOAD_SUCCESS = 'PROJECT_LOAD_SUCCESS';
export const PROJECT_LOAD_FAILED = 'PROJECT_LOAD_FAILED';
