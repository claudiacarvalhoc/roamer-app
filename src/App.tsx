import React from 'react';
import logo from './logo.svg';
import './App.css';

import { useAppSelector, useAppDispatch } from './app/hooks';
import {
  expand,
  collapse,
  selectAppTitle
} from './appSlice';

export default function App() {
  const title = useAppSelector(selectAppTitle);
  // const dispatch = useAppDispatch();
  return (
    <div className="App">
      <h1>{title}</h1>
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>
          Edit <code>src/App.tsx</code> and save to reload.
        </p>
        <a
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </a>
      </header>
    </div>
  );
}
