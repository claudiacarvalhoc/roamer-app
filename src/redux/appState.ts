import { DefaultRootState } from "react-redux";

export interface AppState extends DefaultRootState {
    expand: boolean;
    projects: ProjectProperties[];
    menuTexts: MenuTexts,
    projectTexts: ProjectTextsProperties,
}

export interface MenuTexts {
    projectText: string;
    expandText: string;
    collapseText: string;
}

export interface ProjectTextsProperties {
    doneText: string;
    baseWordsText: string;
    teamText: string;
    keysText: string;
    qaIssuesText: string;
}

export interface IconType {
    Icon: 'up' | 'down' | 'check' | 'user' | 'rate' | 'camera' | 'book';
}

export interface TranslationSectionsProperties {
    country: string;
    language: string;
    done: number;
    wordsToDo: number;
    unverified: number;
}

export interface ProjectProperties {
    id: number;
    name: string;
    done: number;
    baseWords: number;
    team: number;
    keys: number;
    issues: number;
    icons: string[];
    badges: string[];
    translationSections: TranslationSectionsProperties[];
}
