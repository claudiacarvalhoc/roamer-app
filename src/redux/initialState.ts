import { AppState, MenuTextsState, ProjectState, ProjectTextsState } from "./appState";

export const menuInitialState: MenuTextsState = {
    projectText: 'New project ⇧⌘P',
    expandText: 'Expand all',
    collapseText: 'Collapse all',
};

export const projectsInitialState: ProjectTextsState = {
    doneText: 'DONE',
    baseWordsText: 'BASE WORDS',
    teamText: 'TEAM',
    keysText: 'KEYS',
    qaIssuesText: 'QA ISSUES',
    wordsToDoText: 'WORDS TO DO',
    unverifiedText: 'UNVERIFIED',
    addLanguageText: 'Add language',
};

export const projectInitialState: ProjectState = {
    id: 1,
    name: 'Roamer App (Android SDK test)',
    done: 30,
    baseWords: 160246,
    team: 26,
    keys: 1834,
    issues: 47514,
    icons: ['up', 'down', 'check', 'user', 'rate', 'camera', 'book'],
    badges: [
        {
            name: 'Roamer',
            type: 'orange',
        },
        {
            name: 'iOS',
            type: 'green',
        }
    ],
    translationSections: [{
        id: 1,
        country: 'uk',
        language: 'English',
        done: 20,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 2,
        country: 'uk',
        language: 'English',
        done: 20,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 3,
        country: 'uk',
        language: 'English',
        done: 20,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 4,
        country: 'uk',
        language: 'English',
        done: 20,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 5,
        country: 'ru',
        language: 'Russian',
        done: 45,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 6,
        country: 'ru',
        language: 'Russian',
        done: 45,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 7,
        country: 'ru',
        language: 'Russian',
        done: 45,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 8,
        country: 'ru',
        language: 'Russian',
        done: 45,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 9,
        country: 'it',
        language: 'Italian',
        done: 83,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 10,
        country: 'it',
        language: 'Italian',
        done: 83,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 11,
        country: 'it',
        language: 'Italian',
        done: 83,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 12,
        country: 'it',
        language: 'Italian',
        done: 83,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 13,
        country: 'mk',
        language: 'Macedonian',
        done: 67,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 14,
        country: 'mk',
        language: 'Macedonian',
        done: 67,
        wordsToDo: 16246,
        unverified: 50000,
    }, {
        id: 15,
        country: 'mk',
        language: 'Macedonian',
        done: 67,
        wordsToDo: 16246,
        unverified: 50000,
    }]
};

export const appInitialState: AppState = {
    expand: true,
    projects: [
        projectInitialState
    ],
    MenuTextsState: menuInitialState,
    projectTexts: projectsInitialState,
};

export const newProjectInitialState: ProjectState = {
    id: 0,
    name: 'Roamer App (Android SDK test)',
    done: 0,
    baseWords: 160246,
    team: 26,
    keys: 1834,
    issues: 47514,
    icons: ['up', 'down', 'check', 'user', 'rate', 'camera', 'book'],
    badges: [
        {
            name: 'Roamer',
            type: 'orange',
        },
        {
            name: 'iOS',
            type: 'green',
        }
    ],
    translationSections: [],
};