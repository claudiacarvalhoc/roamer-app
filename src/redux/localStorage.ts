import { AppState } from './appState';
const KEY = 'state';

export const loadState = (): AppState => {
    try {
        const serializedState = localStorage.getItem(KEY);
        return serializedState !== null ? JSON.parse(serializedState) : undefined;
    } catch (err) {
        console.error(`Error loading state from local storage`, err);
        return undefined;
    }
};

export const saveState = (state: AppState) => {
    try {
        const serializedState = JSON.stringify(state);
        localStorage.setItem(KEY, serializedState);
    } catch (err) {
        console.error(`Error saving state on local storage`, err);
    }
};