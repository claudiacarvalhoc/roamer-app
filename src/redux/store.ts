import { AnyAction } from 'redux';
import { AppState } from './appState';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { applyMiddleware, compose, createStore } from 'redux';
import { createLogger } from 'redux-logger';
import { rootReducer } from './reducers';
import thunk from 'redux-thunk';
import { loadState, saveState } from './localStorage';

export type AppThunk<R> = ThunkAction<R, AppState, unknown, AnyAction>;
export type AppDispatch = ThunkDispatch<AppState, unknown, AnyAction>;


export const configureStore = (initialState: AppState) => {
    const logger = createLogger();
    const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
    const middleware = [thunk, logger];

    const preloadedState = loadState() ?? initialState;
    const store = createStore(
        rootReducer,
        {
            app: preloadedState,
        },
        composeEnhancers(applyMiddleware(...middleware)),
    );

    store.subscribe(() => {
        saveState(store.getState().app);
    });

    return store;
};

export default configureStore;
