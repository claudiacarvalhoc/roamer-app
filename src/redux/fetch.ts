import { AppDispatch } from './store';
import { addProjectAction } from './app/actions';
import { newProjectInitialState } from './initialState';

export const createProject = (): ((dispatch: AppDispatch) => Promise<void>) => {
    return dispatch => {
        dispatch(addProjectAction(newProjectInitialState));
        return Promise.resolve();
    };
};