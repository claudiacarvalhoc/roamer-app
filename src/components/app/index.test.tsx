import React from 'react';
import App, { AppProps } from './index';
import Enzyme, { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { appInitialState, projectInitialState } from '../../redux/initialState';
import { ProjectState } from '../../redux/appState';

Enzyme.configure({ adapter: new Adapter() });

const getMountWrapper = (projects: ProjectState[] = []) => {
  const store = configureMockStore()({
    app: {
      ...appInitialState,
      projects
    }
  });
  return mount(
      <Provider store={store}>
          <App />
      </Provider>
  );
};

const getShallowWrapper = (projects: ProjectState[] = []) => {
  const store = configureMockStore()({
    app: {
      ...appInitialState,
      projects
    }
  });
  return shallow(
      <Provider store={store}>
          <App />
      </Provider>
  );
};

describe('<App />', () => {
  it('render correctly when do not have any projects', () => {
    const component = getShallowWrapper();
    expect(toJson(component)).toMatchSnapshot();
  });
  it('render correctly when there are at least one project', () => {
    const component = getShallowWrapper([
      projectInitialState
    ]);
    expect(toJson(component)).toMatchSnapshot();
  });
  it('projects are not loaded when do not have any projects', () => {
    const component = getMountWrapper();
    expect(toJson(component)).toMatchSnapshot();
    const hasProjectsElement = component.find('[data-test="app-projects"]').exists();
    expect(hasProjectsElement).toBeFalsy();
  });
  it('projects are loaded when there are at least one project', () => {
    const component = getMountWrapper([
        projectInitialState
      ]);
    const hasProjectsElement = component.find('[data-test="app-projects"]').exists();
    expect(hasProjectsElement).toBeTruthy();
  });
});
