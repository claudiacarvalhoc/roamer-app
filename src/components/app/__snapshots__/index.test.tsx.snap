// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> projects are not loaded when do not have any projects 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(App)>
    <App
      dispatch={[Function]}
      isExpanded={true}
      projects={Array []}
    >
      <div
        className="app"
      >
        <Connect(Menu)>
          <Menu
            addProject={[Function]}
            collapse={[Function]}
            collapseText="Collapse all"
            expand={[Function]}
            expandText="Expand all"
            projectText="New project ⇧⌘P"
          >
            <div
              className="menu"
            >
              <Button
                data-test="button-new-project"
                onClick={[Function]}
                text="New project ⇧⌘P"
                type="primary"
              >
                <button
                  className="button primary"
                  onClick={[Function]}
                >
                  New project ⇧⌘P
                </button>
              </Button>
              <Button
                data-test="button-expand"
                onClick={[Function]}
                text="Expand all"
                type="secondary"
              >
                <button
                  className="button secondary"
                  onClick={[Function]}
                >
                  Expand all
                </button>
              </Button>
              <Button
                data-test="button-collapse"
                onClick={[Function]}
                text="Collapse all"
                type="secondary"
              >
                <button
                  className="button secondary"
                  onClick={[Function]}
                >
                  Collapse all
                </button>
              </Button>
            </div>
          </Menu>
        </Connect(Menu)>
      </div>
    </App>
  </Connect(App)>
</Provider>
`;

exports[`<App /> render correctly when do not have any projects 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(App) />
</ContextProvider>
`;

exports[`<App /> render correctly when there are at least one project 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(App) />
</ContextProvider>
`;
